File: core/src/ru/hyst329/openfool/Card.java
package ru.hyst329.openfool;

import java.util.Locale;

public class Card {
    private final Suit suit;
    private final Rank rank;

    public Card(Suit suit, Rank rank) {
        this.suit = suit;
        this.rank = rank;
    }

    Suit getSuit() {
        return suit;
    }

    Rank getRank() {
        return rank;
    }

    boolean beats(Card other, Suit trumpSuit) {
        int thisRankValue = (this.rank.getValue() + 11) % 13;
        int otherRankValue = (other.rank.getValue() + 11) % 13;
        if (this.suit == other.suit) {
            return thisRankValue >= otherRankValue;
        } else return this.suit == trumpSuit;
    }

    public String toString() {
        return String.format(Locale.ENGLISH, "%d%c", rank.getValue(),
                suit.name().toLowerCase().charAt(0));
    }
}

File: core/src/ru/hyst329/openfool/Card.kt
package ru.hyst329.openfool

import java.util.Locale

class Card(internal val suit: Suit, internal val rank: Rank) {

    internal fun beats(other: Card, trumpSuit: Suit, deuceBeatsAce: Boolean): Boolean {
        val thisRankValue = (this.rank.value + 11) % 13
        val otherRankValue = (other.rank.value + 11) % 13
        if (this.suit === other.suit) {
            // deuce: (2 + 11) % 13 == 0, ace: (1 + 11) % 13 == 12
            return thisRankValue >= otherRankValue || (deuceBeatsAce && thisRankValue == 0 && otherRankValue == 12)
        } else
            return this.suit === trumpSuit
    }

    override fun toString(): String {
        return "${rank.value}${suit.name.toLowerCase()[0]}"
    }
}
